server/app/repository/ApplicationRepository.java

```diff
@@ -76,27 +77,37 @@ private Application submitApplicationInternal(
               .eq("applicant.id", applicant.id)
               .eq("program.name", program.getProgramDefinition().adminName())
               .findList();
-      Optional<Application> completedApplication = Optional.empty();
-      for (Application application : oldApplications) {
-        // Delete any in-progress drafts, and mark obsolete any old applications.
-        if (application.getLifecycleStage().equals(LifecycleStage.DRAFT)) {
-          application.setLifecycleStage(LifecycleStage.ACTIVE);
-          application.setSubmitTimeToNow();
-          completedApplication = Optional.of(application);
-        } else {
-          application.setLifecycleStage(LifecycleStage.OBSOLETE);
-        }
-        application.save();
+      List<Application> latestDraft =
+          oldApplications.stream()
+              .filter(app -> app.getLifecycleStage().equals(LifecycleStage.DRAFT))
+              .collect(Collectors.toList());
+
+      if (latestDraft.size() > 1) {
+        throw new RuntimeException(
+            String.format(
+                "Found more than one DRAFT application " + "for applicant %d, program %d.",
+                applicant.id, program.id));
       }
+
       Application application =
-          completedApplication.orElse(new Application(applicant, program, LifecycleStage.ACTIVE));
-      application.setSubmitTimeToNow();
+          !latestDraft.isEmpty()
+              ? latestDraft.get(0)
+              : new Application(applicant, program, LifecycleStage.ACTIVE);
```

Comment:
suggestion (blocking): reads a bit easier without the inversion
```suggestion
          latestDraft.isEmpty()
              ? new Application(applicant, program, LifecycleStage.ACTIVE)
              : latestDraft.get(0);
```